mllib/src/main/scala/org/apache/spark/ml/feature/StringIndexer.scala

```diff
@@ -151,4 +153,77 @@ class StringIndexerModel private[ml] (
     val copied = new StringIndexerModel(uid, labels)
     copyValues(copied, extra)
   }
+
+  /**
+   * Return a model to perform the inverse transformation.
+   * Note: By default we keep the original columns during this transformation, so the inverse
+   * should only be used on new columns such as predicted labels.
+   */
+  def invert(inputCol: String, outputCol: String): StringIndexerInverseTransformer = {
+    val labelsCol: String = $(this.outputCol)
+    new StringIndexerInverseTransformer(labelsCol)
+      .setInputCol(inputCol)
+      .setOutputCol(outputCol)
+  }
+}
+
+/**
+ * :: Experimental ::
+ * Transform a provided column back to the original input types using the metadata on the
+ * labelsCol. Note: By default we keep the original columns during this transformation,
+ * so the inverse should only be used on new columns such as predicted labels.
+ */
+@Experimental
+class StringIndexerInverseTransformer private[ml] (
+  override val uid: String,
+  val labelsCol: String) extends Transformer
```

Comment:
I'd prefer using an array of labels since I could imagine users also using this transformer for other purposes, such as taking a UCI dataset and converting the categorical indices to strings based on the dataset description.